### PR TITLE
Add persistent batching and minimization to differential tests

### LINKED ISSUE #354: Add persistent batching, minimization, and corpora to differential testing
> This issue was generated by GitHub Copilot after an AI-assisted repository review.

## Context

The differential test currently generates a fixed-seed sample, processes inputs sequentially, and starts a new Lean process for every case. Failures are printed but are not automatically minimized or persisted as regression fixtures. This limits the number of parser states explored per CI minute and makes failures harder to reproduce and maintain.

## Proposed change

- Add a long-lived, framed protocol to the Lean CLI for processing many base64-encoded inputs per process.
- Execute comparisons through a bounded worker pool.
- Record the seed, generator, case index, escape character, and exact replay command for each failure.
- Minimize mismatches with FsCheck shrinkers where available and structure-aware simplification otherwise.
- Require minimization to preserve the same mismatch category rather than merely producing any parse failure.
- Store minimized failures in a checked-in local regression corpus.
- Run the regression corpus before generated cases.
- Keep pull-request runs reproducible while using rotating deterministic seeds in a scheduled workflow.

## Acceptance criteria

- Differential comparison does not launch a new Lean process for every input.
- Concurrency is bounded and deterministic failures remain replayable.
- Every mismatch reports a one-command reproduction.
- Minimized failures can be promoted to a checked-in regression corpus.
- CI runs the regression corpus before randomized generation.
- Scheduled runs explore seeds beyond the pull-request default.

## Dependencies

- **Blocking dependencies:** None.
- **Coordinates with:** #358; minimized local regressions and the pinned upstream corpus remain separate inputs but should share the same differential-test execution path.

### PR DESCRIPTION
**Pull request by mthalman**
The differential test harness currently starts a Lean process per input and does not retain failures for deterministic replay. This makes sustained fuzzing expensive and leaves newly discovered parser differences difficult to reproduce and reduce.

This change adds:

- A framed batch protocol and bounded pool of persistent Lean parser workers, with stable result ordering and explicit parser-versus-infrastructure error classification.
- Deterministic FsCheck generation metadata, exact base64 replay commands, and category-preserving failure minimization with bounded, sampled candidate generation.
- A checked-in regression corpus that runs before generated cases, plus explicit atomic promotion of minimized failures.
- Fixed-seed pull request coverage and a scheduled workflow using deterministic daily rotating seeds.
- Infrastructure tests covering concurrency, cancellation, protocol failures, replay portability, minimization, and corpus validation.

Validation included the full .NET test suite, the Lean differential CLI build, and multi-worker differential runs against generated inputs.

Fixes: #354

### CODE REVIEW

File: src/Valleysoft.DockerfileModel.DiffTest/DiffTestRunner.cs [hunk 1]
```diff
@@ -1,237 +1,196 @@
...
         {
-            // Workaround for #259: VOLUME [] crashes C# parser (empty exec-form array)
-            if (IsKnownCrash(instructionType, input))
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (IsKnownCrash(testCase.InstructionType, testCase.Input))
             {
-                return new DiffResult(instructionType, input, "", "", true);
+                return new DiffResult(testCase, "", "", DiffOutcomeKind.Match);
             }
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel.DiffTest/DiffTestRunner.cs:120]: Fixed. The known `VOLUME []` workaround now invokes Lean before returning `Match`. A Lean parse rejection remains a shared rejection, while startup, protocol, and other infrastructure failures are returned as `InfrastructureError`. I added a regression test proving the known-crash path preserves a Lean infrastructure failure.

_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/Valleysoft.DockerfileModel.DiffTest/LeanProcessWorker.cs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+                responseId != requestId)
+            {
+                throw new LeanInfrastructureException(
+                    $"Malformed Lean response frame for request {requestId}: '{response}'.");
+            }
+
+            string body;
+            try
+            {
+                body = Encoding.UTF8.GetString(Convert.FromBase64String(fields[2]));
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel.DiffTest/LeanProcessWorker.cs]: Added child-process integration coverage for the worker protocol. The tests launch a native fake process, validate the exact request frame (ID, escape code, and base64 input), and exercise `ok`, `parse-error`, `error`, malformed-frame, and EOF responses through `LeanProcessWorker`. The internal constructor only supplies the executable/argument seam; production still starts the Lean CLI with `--batch`.

_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

File: src/Valleysoft.DockerfileModel.DiffTest/ReplayCommand.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+                directory = directory.Parent;
+            }
+        }
+
+        return null;
+    }
+
+    private static string Quote(string value) =>
+        "\"" + value.Replace("\"", "\\\"") + "\"";
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Valleysoft.DockerfileModel.DiffTest/ReplayCommand.cs]: Fixed. Replay arguments now use host-aware single-quote escaping: PowerShell apostrophes are doubled on Windows, and POSIX apostrophes use the standard `'"'"'` sequence. Project, assembly, Lean CLI, instruction, and base64 arguments all use the same quoting helper, with tests covering `$`, backticks, backslashes, and apostrophes.

_Auto-resolved by the [GitHub Copilot app](https://github.com/github/app)._

**Comment by mthalman** [comment]
Addressed the remaining items from review https://github.com/mthalman/DockerfileModel/pull/374#pullrequestreview-5183155761 in commit 9efa6ae:

- Regression fixtures now use strict UTF-8 decoding and reject malformed bytes as invalid fixture data.
- Lean response bodies now use strict UTF-8 decoding, classifying malformed bytes as infrastructure errors.
- stderr collection after stdout EOF is bounded to prevent a child retaining stderr from hanging the worker.
- InputGenerator documentation now describes DiffCase metadata and the instruction plus edge-case generator distribution.

Added focused regressions for malformed fixture UTF-8, malformed worker UTF-8, and bounded stderr draining. The full 1,384-test suite and a 1,000-case four-worker differential run pass.
